helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -413,6 +413,31 @@ public void addRoutingTableChangeListener(
     _routingTableChangeListenerMap.put(routingTableChangeListener, new ListenerContext(context));
     logger.info("Attach RoutingTableProviderChangeListener {}",
         routingTableChangeListener.getClass().getName());
+    if (_sourceDataTypeMap.isEmpty()) {
```

Comment:
This is an initial call when adding listener. The only concern is returning empty data should be a good signal for user? This changes behavior. Some users are relying on the snapshot content. Shall we have a different API that give the option to let user choose it?

Reply:
I see your concern. There're several customers that wanted this feature. However, it's possible to cause unnecessary callback to other customers. Maybe we have a boolean value in the function, and let customers to determine their usage?

Reply:
Let's not change the API. The other customers can still use this API. But let's have another API and old API call the new API with input value as false (for first time refresh) to guarantee the backward compatible.

Reply:
So the users that would like the initial callback will explicitly call the new API?

Reply:
FYI, in Helix Manager, adding the listener will cause the listener to get a "free" callback when the callback handler is init. So the requirement here is doing the same for router.

Given that saying, if the routing table has not been updated (maybe during the reading) then we return an empty snapshot. In this case, if the router still notifies the listener once the update is done, then we are still good. However, based on the code here, I think we might have a race condition that this notification could be sent later than the proper one. Then user would get the wrong information.
So we need a concurrency control here.

Reply:
> So the users that would like the initial callback will explicitly call the new API?

Yes.